cli/internal/cmd/init.go

```diff
@@ -195,12 +196,45 @@ func (i *initCmd) initialize(cmd *cobra.Command, newDialer func(validator *cloud
 	return nil
 }
 
-func (i *initCmd) initCall(ctx context.Context, dialer grpcDialer, ip string, req *initproto.InitRequest) (*initproto.InitResponse, error) {
+func (i *initCmd) initCall(ctx context.Context, dialer grpcDialer, ip string, spinner spinnerInterf, req *initproto.InitRequest) (*initproto.InitResponse, error) {
+	initCtx, initCtxCancel := context.WithCancel(ctx)
+	defer initCtxCancel()
+
+	// This goroutine tracks if we established a gRPC connection successfully.
+	// If yes, it switches the spinner to "Initializing cluster".
+	// If it happens multiple times, we cancel the context and abort.
+	// This usually means we lost the connection and opened a new session (which will likely fail).
+	var connectedTwice bool
```

Comment:
```suggestion
	var connectedTwice bool
```

I think what you should to do here is really just use a single variable (`connected`), to check if a connection was established at least once.
We don't even want to try the second time.
So `connected` starts with the value `false`, if we get a signal from `connReadyChan`, we set `connected` to true and in the error handling below we should check if `connected` is true + we received an error and then return an unretriable error.

Reply:
I mainly used the second variable to differentiate the context was canceled from some other circumstance (e.g. pressing CTRL+C) during a "normal" connection (since I believe it is inherited from Cobra).

Reply:
Has been superseded by just using `context.CancelWithCause`.